api: refactoring

FaiyazMujawar opened the pull request:
Fixes #49 & #63

In this PR,

- Implemented functionality to send proper message to blocked users. The API now sends the following response when a blocked user tries to access any API route


```JSON
status: 403

{
  "code": "FORBIDDEN",
  "result": "FAILURE",
  "message": "You have been blocked by the system for repeated violations of the Code of Conduct"
}
```

- Refactored the codebase to the match coding best practices as much as possible.
- Added code documentation wherever necessary.
- Seperated all the data validation functions into a seperate folder `/misc/validation/`.
- Made the following changes in the codebase to change from `student-grievance-system` to `generic-grievance-system` :

  - models/roles.js
    - `student` &#x2192; `user`
    - `student_moderator` &#x2192; `moderator`
  - models/User.js
    - `zprn` &#x2192; `userId`
  - models/Issue.js
    - `scope` earlier accepted `INSTITUTE/DEPARTMENT` as values. Now accepts `ORGANIZATION/DEPARTMENT`.

FaiyazMujawar committed:
api: refactoring

- Refactored the codebase to the match coding best practices as much as possible.
- Added code documentation wherever necessary.
- Seperated all the data validation functions into a seperate folder `/misc/validation/`.
- Made the following changes in the codebase to change from `student-grievance-system` to `generic-grievance-system` :

  - models/roles.js
    - `student` -> `user`
    - `student_moderator` -> `moderator`
  - models/User.js
    - `zprn` -> `userId`
  - models/Issue.js
    - `scope` earlier accepted `INSTITUTE/DEPARTMENT` as values. Now accepts `ORGANIZATION/DEPARTMENT`.

drigio commented on controllers/IssueController.js:
```diff
@@ -1,35 +1,42 @@
  * @param {Response} res Response Object
  */
-const getAllIssues = async (req, res) => {
+const getAllIssuesController = async function (req, res) {
```
why change arrow function to function keyword?
Edit : Ignore, Its been suggested in Airbnb js style guide.

drigio commented on services/UserService.js:
```diff
@@ -3,125 +3,102 @@ const randomBytes = require("crypto").randomBytes;
-   * The fields that aren't to be updated should be left undefined
+  /**
+   * Updates the specified with provided details
+   * @param {Object} newUser Object containing updated user details
```
We need to specify what `newUser` object will need as its keys.

FaiyazMujawar replied:
It can contain any of the fields in the original document, if they need to be updated. We cannot specify what fields to update.

drigio commented on services/UserService.js:
```diff
@@ -152,6 +129,17 @@ class UserService {
   hasAccessTo = (act, resource) => {
     return async (req, res, next) => {
       try {
+        if (req.user.isDisabled) {
```
Is this check added here for the case when :
User is already logged in but has been disabled later because of violations and the user should not be able to proceed with further functionalities even though he has a valid auth token?

FaiyazMujawar replied:
Yes. You got it right. A blocked user must be restricted from accessing any if the API endpoints, not just the `auth` routes. Instead of checking this condition in every controller, I thought it would be better just to write it in the midlleware passed to all the routes, `DRY` philosophy! 😆

drigio reviewed (commented):
Everything else looks great! Just a few questions.
Will test manually once before merging in. Also unit tests need to be written for the same asap! 
The code coverage report will tell us which functions remain without tests. The code coverage must've dropped significantly.
After the unit tests we need to quickly move on to integration tests so that we will not need to test the functionality every time manually. Also I feel the code is now significantly big and that we should focus now on integration tests and setting up of fake db.

FaiyazMujawar committed:
auth: implemented proper message to blocked users

Implemented functionality to send proper message to blocked users. The API now sends the following response when a blocked user tries to access any API route

```JSON
status: 403

{
  "code": "FORBIDDEN",
  "result": "FAILURE",
  "message": "You have been blocked by the system for repeated violations of the Code of Conduct"
}
```
Also, updated documentation

drigio commented:
How about we add validations in `helpers` rather than `misc`. It feels more natural and widely used term?

FaiyazMujawar commented:
Both ways it's good, according to me😅 You say what to do.
